src/vm/analysis/trait_checker/post_type_check.rs

```diff
@@ -0,0 +1,80 @@
+use std::collections::HashMap;
+
+use vm::analysis::types::{ContractAnalysis, AnalysisPass};
+use vm::analysis::AnalysisDatabase;
+use vm::analysis::errors::{CheckResult, CheckError, CheckErrors};
+use vm::representations::{SymbolicExpression, ClarityName};
+use vm::representations::SymbolicExpressionType::{AtomValue, Atom, List, LiteralValue};
+use vm::types::{Value, TraitIdentifier, TypeSignature, FunctionType};
+use vm::functions::NativeFunctions;
+use vm::functions::define::{DefineFunctions, DefineFunctionsParsed};
+
+pub struct PostTypeCheckingTraitChecker {
+}
+
+impl AnalysisPass for PostTypeCheckingTraitChecker {
+
+    fn run_pass(contract_analysis: &mut ContractAnalysis, analysis_db: &mut AnalysisDatabase) -> CheckResult<()> {
+        let mut command = PostTypeCheckingTraitChecker::new();
+        command.run(contract_analysis, analysis_db)?;
+        Ok(())
+    }
+}
+
+impl PostTypeCheckingTraitChecker {
+
+    fn new() -> Self {
+        Self {
+        }
+    }
+
+    pub fn run(&mut self, contract_analysis: &mut ContractAnalysis, analysis_db: &mut AnalysisDatabase) -> CheckResult<()> {
+    
+        for trait_identifier in &contract_analysis.implemented_traits {
+
+            let trait_name = trait_identifier.name.to_string();
+            let contract_defining_trait = analysis_db.load_contract(&trait_identifier.contract_identifier)
+                .ok_or(CheckErrors::TraitReferenceUnknown(trait_identifier.name.to_string()))?;
+            let trait_sig = contract_defining_trait.get_defined_trait(&trait_name)
+                .ok_or(CheckErrors::TraitReferenceUnknown(trait_identifier.name.to_string()))?;
```

Comment:
What happens if I `impl` a trait that I also define in the same contract?  Would this pass reject such an `impl`?

Reply:
Yes it would. Adding a test case here - https://github.com/blockstack/blockstack-core/pull/1246/commits/e07541934188bdd2010a750e1f5d91d0b2918993.
Ending up passing the contract that you want to dynamically call as an argument sounds sketchy to me, but open to discuss about it.

Reply:
I'm fine with either design, just as long as it's documented and tested :+1:  I commented because I wasn't sure what the intended design was.